src/framework-ondata.js

```
@@ -1,18 +1,164 @@
 
-     return;
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Update the input state
+     var inputState = (payload.vnid) ? payload : stateFactory( vnid, payload );
+     vni.inputStates( portName, socketIndex, inputState );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( vni ) ) { 
+          
+        // Save vars we will need in the promise where the context is different
+        var outputPorts = this.nodeInstance.outPorts;
+
+        if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+        } else { 
+
+            // save current error & output state LMs to compare after updater runs
+            var lastOutputLm = vni.outputState().lm;
+            var lastErrorLm = vni.errorState().lm;
+            var wrapper = this.nodeInstance.wrapper;
+
+            new Promise(function( resolve ) { 
+                resolve( wrapper.fRunUpdater.call( vni ) );
+
+            }).then( function() { 
+                // If not new output or error state data, send it downstream or log it to the console
+                handlePortOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                handlePortOutput( outputPorts.error,  lastErrorLm, vni.errorState() );
+
+            }, function( rejected ) { 
+    
+                console.error( rejected );
+
+                // fRunUpdater failed - is this a new error? 
+                var errorState = vni.errorState();
+
+                if ( lastErrorLm !== errorState.lm  ) { 
+
+                    // fRunUpdater or updater recorded a new lm (i.e., new error) - send it on
+                    handlePortOutput( outputPorts.error, lastErrorLm, errorState );
```

If fRunUpdater or updater recorded a new lm (i.e., new error) *and* throw a rejected error, they are likely different and the system should not swallow the second error, but pass it along as well.